storage/include/sparse_merkle/internal_node.h

```diff
@@ -0,0 +1,280 @@
+    });
+  }
+
+ private:
+  // Return true if the index does not contain a child.
+  bool is_empty(size_t index) { return !children_.at(index).has_value(); }
+
+  // Return true if the index has a child that contains an internal node.
+  bool is_internal(size_t index) {
+    if (auto& child = children_.at(index)) {
+      return std::holds_alternative<InternalChild>(child.value());
+    }
+    return false;
+  }
+
+  // Attempt to insert into an empty child.
+  //
+  // Return InsertComplete if the node was empty.
+  // Otherwise return std::nullopt.
+  std::optional<InsertResult> insert_into_empty_child(size_t index, LeafChild& child);
+
+  // Attempt to overwrite a matching key at index.
+  //
+  // Return InsertComplete if the key was overwritten.
+  // Otherwise return std::nullopt.
+  std::optional<InsertResult> overwrite(size_t index, LeafChild& child);
+
+  // There has been a collision of leaf keys that do not match.
+  //
+  // One or more parents must be created to store them.
+  //
+  // `depth` is the current depth of this BatchedInternalNode in ths sparse merkle
+  // tree.
+  //
+  // Return either CreateNewBatchedInternalNodes or InsertComplete.
+  InsertResult split_node(size_t index, LeafChild& child, size_t depth);
+
+  // Return the hash of a given index.
+  const Hash& get_hash(size_t index) const {
+    auto& child = children_.at(index);
+    if (!child) {
+      return PLACEHOLDER_HASH;
+    }
+    return std::visit([](auto&& arg) -> const Hash& { return arg.hash; }, child.value());
+  }
+
+  // Walk the tree from `index` up to the root of this BatchedInternalNode to
+  // compute the root hash.
+  //
+  // Update the hashes and versions of nodes along the way.
+  void update_hashes(size_t index, Version version);
+
+  // Return the the height of the node at the given index.
+  size_t height(size_t index) {
+    if (index == 0) return 4;
+    if (index < 3) return 3;
+    if (index < 7) return 2;
+    if (index < 15) return 1;
+    return 0;
+  }
+
+  // Return the index of the parent node, given the index of the child node.
+  //
+  // Return std::nullopt if the index points to the root of this
+  // BatchedInternalNode.
+  std::optional<size_t> parent_index(size_t index) {
+    if (index == 0) return std::nullopt;
+    return (index - 1) / 2;
+  }
+
+  // Return the index of the left child of a node at a given index.
+  size_t left_child_index(size_t index) { return 2 * index + 1; }
+
+  // Return the index of the right child of a node at a given index.
+  size_t right_child_index(size_t index) { return 2 * index + 2; }
+
+  // Is this node a left child?
+  bool is_left_child(size_t index) const { return index % 2 != 0; }
+
+  // All internal and leaf nodes are stored in this array, starting from the
+  // root and proceeding level by level from left to right.
+  std::array<std::optional<Child>, MAX_CHILDREN> children_;
+};
+
+}  // namespace sparse_merkle
+}  // namespace storage
+}  // namespace concord
```

Comment:
BTW, why do we have .h and not .hpp files?

Reply:
All the storage code has .h files. I just kept to that convention.